include/xtensor/xoptional_assembly_storage.hpp

```diff
@@ -0,0 +1,551 @@
+    {
+        return lhs.value() <= rhs.value() && lhs.has_value() == rhs.has_value();
+    }
+
+    template <class VE, class FE>
+    bool operator>(const xoptional_assembly_storage<VE, FE>& lhs, const xoptional_assembly_storage<VE, FE>& rhs)
+    {
+        return lhs.value() > rhs.value() && lhs.has_value() == rhs.has_value();
+    }
+
+    template <class VE, class FE>
+    bool operator>=(const xoptional_assembly_storage<VE, FE>& lhs, const xoptional_assembly_storage<VE, FE>& rhs)
+    {
+        return lhs.value() >= rhs.value() && lhs.has_value() == rhs.has_value();
+    }
+
+    template <class VE, class FE>
+    void swap(xoptional_assembly_storage<VE, FE>& lhs, xoptional_assembly_storage<VE, FE>& rhs) noexcept
+    {
+        lhs.swap(rhs);
+    }
+
+    /******************************************************
+     * xoptional_assembly_storage_iterator implementation *
+     ******************************************************/
+
+    template <class VE, class FE, bool C>
+    inline xoptional_assembly_storage_iterator<VE, FE, C>::xoptional_assembly_storage_iterator(value_iterator value_it, flag_iterator flag_it)
+        : m_value_it(value_it), m_flag_it(flag_it)
+    {
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator++() -> self_type&
+    {
+        ++m_value_it;
+        ++m_flag_it;
+        return *this;
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator--() -> self_type&
+    {
+        --m_value_it;
+        --m_flag_it;
+        return *this;
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator+=(difference_type n) -> self_type&
+    {
+        m_value_it += n;
+        m_flag_it += n;
+        return *this;
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator-=(difference_type n) -> self_type&
+    {
+        m_value_it -= n;
+        m_flag_it -= n;
+        return *this;
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator-(const self_type& rhs) const -> difference_type
+    {
+        return m_value_it - rhs.m_value_it;
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator*() const -> reference
+    {
+        return reference(*m_value_it, *m_flag_it);
+    }
+
+    template <class VE, class FE, bool C>
+    inline auto xoptional_assembly_storage_iterator<VE, FE, C>::operator->() const -> pointer
+    {
+        return &(this->operator*());
+    }
+
+    template <class VE, class FE, bool C>
+    inline bool xoptional_assembly_storage_iterator<VE, FE, C>::operator==(const self_type& rhs) const
+    {
+        return m_value_it == rhs.m_value_it && m_flag_it == rhs.m_flag_it;
```

Comment:
isn't it enough to compare one of the two?

Reply:
it should be enough yes